File: src/main/java/ru/ispras/microtesk/test/template/Block.java
/*
 * Copyright (c) 2014 ISPRAS (www.ispras.ru)
 * 
 * Institute for System Programming of Russian Academy of Sciences
 * 
 * 25 Alexander Solzhenitsyn st. Moscow 109004 Russia
 * 
 * All rights reserved.
 * 
 * Block.java, Aug 27, 2014 12:24:07 PM Andrei Tatarnikov
 * 
 * Licensed under the Apache License, Version 2.0 (the "License"); you may not use this file except
 * in compliance with the License. You may obtain a copy of the License at
 * 
 * http://www.apache.org/licenses/LICENSE-2.0
 * 
 * Unless required by applicable law or agreed to in writing, software distributed under the License
 * is distributed on an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express
 * or implied. See the License for the specific language governing permissions and limitations under
 * the License.
 */

package ru.ispras.microtesk.test.template;

import java.util.Collections;
import java.util.Map;

import ru.ispras.microtesk.test.sequence.Sequence;
import ru.ispras.microtesk.test.sequence.iterator.IIterator;

public final class Block {
  private final BlockId blockId;
  private final IIterator<Sequence<Call>> iterator;
  private final Map<String, Object> attributes;

  Block(BlockId blockId, IIterator<Sequence<Call>> iterator, Map<String, Object> attributes) {
    if (null == blockId) {
      throw new NullPointerException();
    }

    if (null == iterator) {
      throw new NullPointerException();
    }

    if (null == attributes) {
      throw new NullPointerException();
    }

    this.blockId = blockId;
    this.iterator = iterator;
    this.attributes = attributes;
  }

  Block(BlockId blockId, IIterator<Sequence<Call>> iterator) {
    this(blockId, iterator, Collections.<String, Object>emptyMap());
  }

  public IIterator<Sequence<Call>> getIterator() {
    return iterator;
  }

  public Object getAttribute(String name) {
    return attributes.get(name);
  }

  public BlockId getBlockId() {
    return blockId;
  }
}

File: src/main/arch/demo/x86/templates/block.rb
#
# Copyright 2017 ISP RAS (http://www.ispras.ru)
#
# Licensed under the Apache License, Version 2.0 (the "License");
# you may not use this file except in compliance with the License.
# You may obtain a copy of the License at
#
#   http://www.apache.org/licenses/LICENSE-2.0
#
# Unless required by applicable law or agreed to in writing, software
# distributed under the License is distributed on an "AS IS" BASIS,
# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
# See the License for the specific language governing permissions and
# limitations under the License.
#

require_relative 'x86_base'

#
# Description:
#
# This test template demonstrates how to use instruction blocks.
#
class BlockTemplate < X86BaseTemplate

  def run
    # Produces a single test case that consists of three instructions
    sequence {
      mov_r16r16 ax, bx
      sub_r16r16 cx, dx
      add_r16r16 r16(_), r16(_)
    }.run

    # Atomic sequence. Works as sequence in this context.
    atomic {
      mov_r16r16 ax, bx
      add_r16r16 cx, dx
      sub_r16r16 r16(_), r16(_)
    }.run

    # Produces three test cases each consisting of one instruction
    iterate {
      mov_r16r16 ax, bx
      sub_r16r16 cx, dx
      add_r16r16 r16(_), r16(_)
    }.run

    # Produces four test cases consisting of two instructions
    # (Cartesian product composed in a random order)
    block(:combinator => 'product', :compositor => 'random') {
      iterate {
        sub_r16r16 cx, dx
        add_r16r16 ax, bx
      }

      iterate {
        mov_r16r16 ax, bx
        sub_r16r16 r16(_), r16(_)
      }
    }.run

    # Merges two sequnces in random fashion. Atomic sequences are unmodifiable.
    block(:combinator => 'diagonal', :compositor => 'random', :obfuscator => 'random') {
      sequence {
        sub_r16r16 bx, ax
        or_r16r16 cx, dx
      }

      atomic {
        prologue { comment 'Atomic starts' }
        epilogue { comment 'Atomic ends' }

        and_r16r16 r16(_), r16(_)
      }
    }.run
  end

end
